### DISCUSSION TITLE
Stopping all executions related to an AHK instance

### DISCUSSION THREAD

**Discussion by cranium481** [spyoungtech/ahk#205] [Q&A]
I have an AHK instance which is being used to run AHK commands somewhere. 
I also use the same AHK instance to listen for hotkeys. 
Is there a way for a hotkey to cause all uses of the AHK instance to stop, including the commands not called by hotkeys? 
Something like the ExitApp or Exit functions in the actual AHK
Thanks for any responses in advance

**Comment by spyoungtech** [comment]
I suppose it's possible for a hotkey to trigger a Python function that forcibly terminates all AHK processes. Maybe something like this?

```python
def kill_ahk():
    subprocess.run(['taskkill', '/F', '/IM', 'AutoHotkey.exe']) # or similar? Process image may be different.

# win+k to kill all autohotkey processes
ahk.add_hotkey('#k', callback=kill_ahk)
```
But that will obviously cause problems if subsequent commands are issued.

But all processes are _supposed_ to be cleaned up at exit (`kill` is [registered at exit](https://docs.python.org/3/library/atexit.html#module-atexit)). So if your python script creates an AHK process with the public API, when your script process dies, it _should_ also cleanup any running AHK processes that it created. If this doesn't happen in normal circumstances I'd consider it a bug. Some abnormal circumstances may also result in processes not being cleaned up, like if `sys._exit` is used instead of `sys.exit` or `SystemExit` or other normal Python script exit conditions.

There's currently no graceful shutdown command and it's something I'd like to add to the API. Right now, there is just an (internal) api to kill the process(es).

For example, to kill the main process used to run non-hotkey/hotstring command:
```python
ahk._transport._proc.kill()
```
Hotkeys use a separate process that can also be killed similarly (I think this works).
```python
ahk._transport._hotkey_transport._proc.kill()
```
You can also see some of this in the unittest teardowns, like as [seen here](https://github.com/spyoungtech/ahk/blob/d07c45bf91769528316bc081310c6efff9e5def8/tests/_async/test_hotkeys.py#L25-L26).


Of course, these internal APIs are subject to change and I wouldn't recommend relying on them. But I think the ideal solution would be to have a public API to gracefully (and maybe optionally, forcefully) shutdown the process(es) used by `ahk`.

**Reply by cranium481** [reply]
Thanks for the response! Look forward to an inbuilt shutdown in future releases
